### ISSUE TITLE
Embargoes and Leases do not automatically expire

### ISSUE THREAD

**Issue by chrisdaaz** [samvera/hyrax#2657]
### Descriptive summary

Embargoes and leases do not auto-expire, as currently implemented. You have to expire them in the UI. (And I don't believe the UI is fully baked yet.)

### Expected behavior

Embargoes and leases should auto-expire according to the deposit settings.

### Related work

This issue is moved from: https://github.com/curationexperts/nurax/issues/90

**Comment by mjgiarlo** [comment]
I'm not sure they are currently designed to auto-expire. AFAIK, they've always required human intervention. This may not be a bug but a change in feature behavior. We should get folks' input on this change, particularly the @samvera/hyrax-repo-managers

**Comment by newmanld** [comment]
My input is that most repo-managers would like notifications sent to users of embargoes/leases about to expire (frequency to be configurable?) combined with an auto-expiration function.  I suppose some might want to turn auto-expiration off, if they have a repo manager with the time to do this manually, but I think most will not.

**Comment by vantuyls** [comment]
i agree with @newmanld's assessment. though i would strongly encourage expiration (auto or manual) to be configurable/switchable so that people can do with this what they will. I've heard from a few folks (maybe @mark-dce) who have indicated that their use cases want manual removal of embargoes.

**Comment by mark-dce** [comment]
If you have really sensitive stuff embargoed, the most conservative approach is to notify staff that an embargo has expired and then have the embargo manually lifted once the expiration has been verified.  This is definitely the less frequent case, but we've seen it at for at least two institutions and I think it's pretty important to be able to support this use case without too much extra work.  A configuration option would work.  At present, we've used a cron job running a rake task for those institutions wanting auto-expirations - would documenting that solution be a reasonable response to this ticket?

**Comment by no-reply** [comment]
I'd like to break this ticket down to two separate tickets:

 - [ ] Implement embargo/lease expiry notifications.
   - [ ] Notify (who???) when an embargo is set to expire N days from now.
   - [ ] Make notifications configurable.
  - [ ] Ship an embargo/lease management rake task and document its use in the Hyrax Management Guide.
    - [ ] Task sends notifications, and;
    - [ ] optionally (by argument): Lifts expired embargos and leases.

As far as I can see, embargo and lease is completely undocumented in the Management guide to date. The embargo APIs are less than ergonomic, and under documented. I've seen at least one custom task in the wild that uses `deactivate!` to try to lift embargoes, to no effect.

This is tangentially connected to #3058. Throughout the stack, "deactivate" for embargo and lease *does not* include changing the permissions on the related object. In the UI, permissions change only if the embargo is already expired; in the API, permission changes are a separate method call from deactivation.

**Comment by no-reply** [comment]
Yeah, I think a cron job is the natural approach for this.

**Comment by chrisdaaz** [comment]
+1 to splitting this ticket in two.

does shipping a rake task / chron job mean that local implementations would require a developer to configure this? would it be too difficult to make auto-expiration a configurable setting in the administrative dashboard?

**Comment by no-reply** [comment]
> does shipping a rake task / chron job mean that local implementations would require a developer to configure this? would it be too difficult to make auto-expiration a configurable setting in the administrative dashboard?

This would be a deployment/production consideration, rather than strictly a "developer" one. The trouble with auto-expiration as a part of the application is that it would involve some kind of recurring automated check for expired embargoes/leases, which isn't a great fit for a running web application.

**Comment by mark-dce** [comment]
We're already using sidekiq to manage delayed jobs.  Would it be worthwhile to look into using a sidekiq add-in to provide cron-like functionality via sidekiq (which we have to run anyway)?

REFERENCES
https://github.com/mperham/sidekiq/wiki/Scheduled-Jobs
https://github.com/moove-it/sidekiq-scheduler
https://rubygems.org/search?utf8=✓&query=sidekiq-scheduler

**Comment by no-reply** [comment]
Hyrax now provides `EmbargoExpiryJob` and `LeaseExpiryJob`, which should support sidekiq scheduling (for sidekiq users) or other kinds of ActiveJob engine approaches to this problem.

#4806 proposes rake tasks that will also do the job, and schedules cron jobs in the Helm chart.

is it okay to close this with those resolutions or do we need a pure ruby solution for scheduled embargo lifting?

**Comment by mark-dce** [comment]
**TL/DR**
Closing this ticket seems like a sane course of action to me.

**DETAILED THOUGHTS**
From my perspective, I think providing rake tasks that can be run to trigger expirations and sample configs to run these in a sidekiq or cron type schedule definitely covers the first half of the ask in this ticket.  

There's also a request here for some kind of notification task that would let the system notify users when their embargo was about to expire within the next N days/weeks (i.e. some kind of configurable timeframe with a sensible default).  It doesn't look like that's covered in the PRs linked here.

It feels reasonable to split the request for a notification process out into a separate ticket AND/OR explicitly state here that kind of feature would need to be a local customization at this time.  Im thinking the notification process is significantly more complex and localized, because it touches on who to notify, how (i.e. do you need a mailer set up), when (what the schedule, how much lead time for notices), etc.   That's a lot to sort out and I agree that it should be handled outside this ticket.
